### PR TITLE
#12 debug page

### LINKED ISSUE #12: Create Debug Page
### Description

We want to have a page that shows every piece of data available. This will allow the engineers to be able to use the dashboard to test certain aspects of the car. 

### Acceptance Criteria

Mocks are filled, all available data is put in and correct. 

### Proposed Solution

new File/ folder for pages
Make Grid for double table
One side has all odd values of data
Other side has all even values of data
Create table cell object with name/value/units(apparently telemetry hub is adding code for units :))
Hook Up Table With Controller/refactor Nero_view to take in different views
Add scrolling with buttons/ page switching
N/A for data that doesnt exist.

### Mocks

image

### LINKED ISSUE #15: Make ner_processing a live submodule of telemetry hub’s ner_processing
### Description

In order to not have to copy and paste over the folder everytime we want to update can messaging, make ner_processing be a git submodule of telemetry hub’s ner_processing. 

### Acceptance Criteria

Is a submodule 

### Proposed Solution

Follow githubs guide for creating a submodule.

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by Peyton-McKee**
## Changes

Factored out home to a page, created debug table, created functions for getting input from mock model.

## Notes

Not sure this is the best way to do this, curious if I should be using a different data structure for the table. Also curious on the best way to get input from model/can...

## Screenshots
Screen Shot 2023-02-11 at 9 45 54 PM

Screen Shot 2023-02-23 at 1 59 34 PM
## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #12  (issue #)

### CODE REVIEW

**Comment by mccauleyma** [comment]
REEEEEE this is making me realize that we probably do need to rethink the dashboard <-> MPU interface. @nwdepatie I think we're gonna need to send button inputs to dashboard and then have dashboard keep track of a vehicle state and send that to MPU

File: nero_view.py [hunk 1]
```diff
@@ -1,211 +1,139 @@
...
 
 
 class NeroView(customtkinter.CTk):
-    def __init__(self, controller) -> None:
+    def __init__(self) -> None:
         super().__init__()
-
-        self.controller = controller
+        self.isLinux = platform.platform()[0:5] == "Linux"
+        self.controller = RaspberryModel() if self.isLinux else MockModel()
```
Comment (anthonybernardi) [inline nero_view.py]: self.model

File: mock_model.py [hunk 2]
```diff
@@ -56,3 +71,33 @@ def get_state_of_charge(self) -> Optional[int]:
 
     def get_lv_battery(self) -> Optional[int]:
         return self.lv_battery
+
+    def get_generic(self, id: int) -> Optional[int]:
+        return None
+
+    def get_debug_table_values(self) -> List[debug.Table_Row_Value]:
+        return [debug.Table_Row_Value(0, "speed", self.mph, "mph"), debug.Table_Row_Value(1, "status", self.status, "bool"), debug.Table_Row_Value(2, "dir", self.dir, "bool"), debug.Table_Row_Value(3, "pack temp", self.pack_temp, "C"), debug.Table_Row_Value(4, "motor temp", self.motor_temp, "C"), debug.Table_Row_Value(5, "state of charge", self.state_of_charge, "%"), debug.Table_Row_Value(6, "lv battery", self.lv_battery, "V")]
```
Comment (anthonybernardi) [inline mock_model.py]: this should be represented and stored as a class

File: pages/debug.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
+from tkinter import Frame
+import customtkinter
+from typing import List
+
+class Table_Row_Frame(Frame):
```
Comment (anthonybernardi) [inline pages/debug.py]: all these classes sohuld start with Debug_

File: raspberry_model.py [hunk 4]
```diff
@@ -67,3 +67,38 @@ def get_state_of_charge(self) -> Optional[int]:
 
     def get_lv_battery(self) -> Optional[int]:
         return self.current_data[63]
+
+    def get_generic(self, id: int) -> Optional[int]:
```
Comment (anthonybernardi) [inline raspberry_model.py]: get_by_id

File: nero_view.py [hunk 5]
```diff
@@ -1,211 +1,139 @@
...
-        self.check_can()
-        self.update()
+        # create the container frame that holds all views
+        container = Frame(self)
+        container.pack(side="top", fill="both", expand=True)
+        container.grid_rowconfigure(0, weight=1)
+        container.grid_columnconfigure(0, weight=1)
+
+        # create the views that the container will hold
+        self.frames = {}
```
Comment (anthonybernardi) [inline nero_view.py]: if we're doing a dictionary, we shouldn't do index and should do names instead

File: nero_view.py [hunk 6]
```diff
@@ -1,211 +1,139 @@
...
     def update_speed(self):
         new_mph: Optional[int] = self.controller.get_mph()
         new_kph: Optional[int] = self.controller.get_kph()
 
         new_mph_text = str(new_mph) if new_mph else "N/A"
         new_kph_text = str(new_kph) if new_kph else "N/A"
 
-        self.mph.configure(text=new_mph_text)
-        self.kph.configure(text=new_kph_text)
+        self.frames[0].mph.configure(text=new_mph_text)
```
Comment (anthonybernardi) [inline nero_view.py]: store the home frame as `self.home_frame` as well as putting it in the dict that way we can directly access it here

File: nero_view.py [hunk 7]
```diff
@@ -1,211 +1,139 @@
...
+        # create the container frame that holds all views
+        container = Frame(self)
+        container.pack(side="top", fill="both", expand=True)
+        container.grid_rowconfigure(0, weight=1)
+        container.grid_columnconfigure(0, weight=1)
+
+        # create the views that the container will hold
+        self.frames = {}
+        index = 0
+        for F in (home.Home, debug.Debug_Table):
```
Comment (anthonybernardi) [inline nero_view.py]: F????? what is F

File: nero_view.py [hunk 8]
```diff
@@ -1,211 +1,139 @@
...
+            frame = F(parent=container, controller=self)
```
Comment (anthonybernardi) [inline nero_view.py]: does it need to be called controller?

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
can you leave documentation especially throughout the debug stuff because just looking at it it's hard to understand

File: mock_model.py (hunk 2 above)
Comment (Peyton-McKee) [reply mock_model.py]: stored as a class or an object?

File: nero_view.py (hunk 8 above)
Comment (Peyton-McKee) [reply nero_view.py]: no

File: nero_view.py (hunk 6 above)
Comment (Peyton-McKee) [reply nero_view.py]: what about that case with the view_index

File: nero_view.py (hunk 7 above)
Comment (Peyton-McKee) [reply nero_view.py]: Frame

File: mock_model.py (hunk 2 above)
Comment (anthonybernardi) [reply mock_model.py]: what's the difference

Comment (anthonybernardi) [reply mock_model.py]: (class)

Comment (Peyton-McKee) [reply mock_model.py]: I think Im misunderstanding what your saying, but how would i access the mph, lv batt and stuff like that from a different class?

**Review by anthonybernardi — APPROVED** [review]
merging cuz it's good enough for now and this pr cannot get bigger

**Pull request merged**